2.3-TapTree.ipynb

```diff
@@ -0,0 +1,500 @@
+    "                                       ts_desc_A,\n",
+    "                                       ts_desc_B,\n",
+    "                                       ts_desc_C)\n",
+    "\n",
+    "# Generate TapTree from descriptor.\n",
+    "taptree = TapTree()\n",
+    "taptree.from_desc(tp_desc)\n",
+    "print(taptree.desc)\n",
+    "\n",
+    "# Compute Taproot Output:\n",
+    "taproot_script, tweak, control_map = taptree.construct()\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 1B - TapTree Construction\n",
+    "\n",
+    "In addition to the TapTree class, we provide Node` and TapLeaf classes to capture internal and leaf nodes in a Taptree.\n",
+    "\n",
+    "![test](images/TapTree2.jpg)\n",
+    "\n",
+    "\n",
+    "* **Class `Node`:**\n",
+    "    * Node child members: \n",
+    "        * `Node.left`, `Node.right`\n",
+    "        * (Left/Right child ordering does not affect taproot tweak derivation)\n",
+    "\n",
+    "* **Class `TapLeaf`:**\n",
+    "    * Construct from descriptor string.\n",
+    "        * `TapLeaf.from_desc(descriptor_string)`\n",
+    "    \n",
+    "#### Example 1.2 - Constructing a Taptree with `Node ` & `TapLeaf`"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Run example 1.1 first.\n",
+    "\n",
+    "# Taproot\n",
+    "taptree = TapTree()\n",
+    "taptree.key = pubkey_internal\n",
+    "\n",
+    "# TapLeafs\n",
+    "tapleaf_A = TapLeaf()\n",
+    "tapleaf_B = TapLeaf()\n",
+    "tapleaf_C = TapLeaf()\n",
+    "tapleaf_A.from_desc(ts_desc_A)\n",
+    "tapleaf_B.from_desc(ts_desc_B)\n",
+    "tapleaf_C.from_desc(ts_desc_C)\n",
+    "\n",
+    "# TapTree\n",
+    "taproot = Node()\n",
+    "tapbranch_AB = Node()\n",
+    "tapbranch_AB.left = tapleaf_A\n",
+    "tapbranch_AB.right = tapleaf_B\n",
+    "taproot.left = tapbranch_AB\n",
+    "taproot.right = tapleaf_C\n",
+    "taptree.root = taproot\n",
+    "\n",
+    "# Check if same construction as with taptree descriptor.\n",
+    "taproot_script2, tweak2, control_map2 = taptree.construct()\n",
+    "\n",
+    "# Compare with example 1.1\n",
+    "assert(taproot_script2 == taproot_script)\n",
+    "print(taptree.desc)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 1C - Huffman Constructor\n",
+    "\n",
+    "A TapTree can also be constructed with a greedy algorithm based on the huffman encoding algorithm. \n",
+    "Simply provide the huffman constructor method with the relative probability weights of each TapLeaf.\n",
+    "\n",
+    "* **A higher weight** \n",
+    "    * Indicates a higher likelihood of execution.\n",
+    "    * Means the script will be placed closer to the root if possible.\n",
+    "    * Results in a smaller inclusion proof and lower spending fees.\n",
+    "\n",
+    "![test](images/TapTree3.jpg)\n",
```

Comment:
This graphic isnt clear to me. Perhaps because I am dense. Heres a few suggestions:

- Is the last graphic intentionally missing the weight at the root? I would have expected 7
- Perhaps keep in the 2 in the second graphic and the 2,4 in the 3 graphic
- I dont understand "connect node pair with lowest weight total" and how it relates to the graphic

Reply:
I've moved this to an optional extra chapter.

Reply:
Tracked here: https://github.com/bitcoinops/taproot-workshop/issues/54